code/game/gamemodes/miniantags/pulsedemon/pulsedemon.dm

```diff
@@ -0,0 +1,841 @@
+	var/charge = 1000
+	/// Maximum quantity of power the demon can hold at once.
+	var/maxcharge = 1000
+	/// Book keeping for objective win conditions.
+	var/charge_drained = 0
+	/// Controls whether the demon will drain power from sources. Toggled by a spell.
+	var/do_drain = TRUE
+	/// Amount of power (in watts) to drain from power sources every Life tick.
+	var/power_drain_rate = 1000
+	/// Maximum value for power_drain_rate based on upgrades.
+	var/max_drain_rate = 1000
+
+	/// Amount of power (in watts) required to regenerate health.
+	var/power_per_regen = 1000
+	/// Amount of health lost per Life tick when the power requirement was not met.
+	var/health_loss_rate = 5
+	/// Amount of health regenerated per Life tick when the power requirement was met.
+	var/health_regen_rate = 3
+	/// Lock health regeneration while this is not 0, decreases by 1 every Life tick.
+	var/regen_lock = 0
+	/// Tracking to prevent multiple EMPs in the same tick from instakilling a demon.
+	var/emp_debounce = FALSE
+
+	/// Controls whether the demon can move outside of cables. Toggled by a spell.
+	var/can_exit_cable = FALSE
+	/// Speed used while moving inside cables.
+	var/inside_cable_speed = -0.5
+	/// Speed used while moving outside cables. Can be upgraded.
+	var/outside_cable_speed = 5
+
+	/// The time it takes to hijack APCs and cyborgs.
+	var/hijack_time = 30 SECONDS
+
+	/// The color of light the demon emits. The range of the light is proportional to charge.
+	var/glow_color = "#bbbb00"
+
+	/// Area being controlled, should be maintained as long as the demon does not move outside a container (APC, object, robot, bot).
+	var/area/controlling_area
+	/// Inhabited cable, only maintained while on top of the cable.
+	var/obj/structure/cable/current_cable
+	/// Inhabited power source, maintained while inside, or while inside its area if it is an APC.
+	var/obj/machinery/power/current_power
+	/// Inhabited item, only items which can be used in rechargers can be hijacked. Only maintained while inside the item.
+	var/obj/item/current_weapon
+	/// Inhabited cyborg, only maintained while inside the cyborg.
+	var/mob/living/silicon/robot/current_robot
+	/// Inhabited bot, only maintained while inside the bot.
+	var/mob/living/simple_animal/bot/current_bot
+
+	/// Delay tracker for movement inside bots.
+	var/bot_movedelay = 0
+	/// A cyborg that has already been hijacked can be re-entered instantly.
+	var/list/hijacked_robots
+
+	/// Images of cables currently being shown on the client.
+	var/list/cable_images = list()
+	/// Images of APCs currently being shown on the client.
+	var/list/apc_images = list()
+	/// List of all previously hijacked APCs.
+	var/list/hijacked_apcs = list()
+	/// Reference to the APC currently being hijacked.
+	var/obj/machinery/power/apc/apc_being_hijacked
+	/// This demon's progressbar helper, used for the hijacking timer on APCs and all charger types (cell, gun, cyborg).
+	var/datum/progressbar_helper/pb_helper
```

Comment:
you are holding a LOT of hard refs here to a lot of things that can easily be destroyed. this will lead to GC issues. can you please make use of UIDs (where it makes sense to)?